File: LeetCode/50.Pow(x,_n)/Mario/50.Pow(x,_n).java
class Solution {
    public double myPow(double x, int n) {
        long m = n;
        return m >= 0 ? helper(x, m) : 1.0 / helper(x, -m);
    }

    private double helper(double x, long n) {
        if (n == 0) {
            return 1.0;
        }
        double y = helper(x, n / 2);
        return n % 2 == 0 ? y * y : y * y * x;
    }
}

File: LeetCode/50.Pow(x,_n)/Mario/50.Pow(x,_n).swift
class Solution {
    func myPow(_ x: Double, _ n: Int) -> Double {
        if n == 0 { return 1.0 }
        if x == 0 { return 0.0 }

        var e: Int = n, base: Double = x
        if e < 0 {
            e = -e
            base = 1.0 / base
        }

        var res = 1.0
        while e > 0 {
            if e & 1 == 1 {
                res *= base
            }

            base *= base
            e >>= 1
        }

        return res
    }
}
